File: src/reaper_mcp/bridge_client.py
"""File-based IPC client for talking to the reaper_bridge.lua ReaScript.

reaper_bridge.lua polls a "requests" directory on every reaper.defer() tick
(REAPER's UI frame rate), so round trips land in roughly one frame
(~16-33ms) without requiring any REAPER extension. This client writes one
JSON file per request, waits for the matching response file to appear, then
cleans both up.

Directory layout (mirrors the Lua side, which derives it from
reaper.GetResourcePath()):
    <bridge_dir>/requests/req_<id>.json
    <bridge_dir>/responses/resp_<id>.json
    <bridge_dir>/heartbeat.txt   (touched every tick the bridge script is alive)
"""

from __future__ import annotations

import itertools
import json
import os
import tempfile
import time
from dataclasses import dataclass, field
from pathlib import Path

HEARTBEAT_STALE_AFTER_SEC = 2.0


class BridgeError(RuntimeError):
    """Raised when the bridge is unreachable or returns an error response."""


class BridgeNotConnected(BridgeError):
    """Raised when the bridge directory/heartbeat indicates the script isn't running."""


def default_bridge_dir() -> Path:
    """Resolve the bridge IPC directory, matching reaper_bridge.lua's own
    reaper.GetResourcePath() + "/Scripts/reaper_mcp_bridge" derivation."""
    override = os.environ.get("REAPER_MCP_BRIDGE_DIR")
    if override:
        return Path(override)

    from .discovery import find_reaper_installs

    installs = find_reaper_installs()
    if installs:
        return Path(installs[0].resource_path) / "Scripts" / "reaper_mcp_bridge"

    # last-resort fallback so callers still get a stable path to report in errors
    return Path.home() / ".reaper_mcp_bridge"


@dataclass
class BridgeConfig:
    bridge_dir: Path = field(default_factory=default_bridge_dir)
    request_timeout: float = 5.0
    poll_interval: float = 0.01


class BridgeClient:
    """Writes request files and polls for response files in the bridge directory."""

    def __init__(self, config: BridgeConfig | None = None):
        self.config = config or BridgeConfig()
        self._id_counter = itertools.count(1)

    @property
    def requests_dir(self) -> Path:
        return self.config.bridge_dir / "requests"

    @property
    def responses_dir(self) -> Path:
        return self.config.bridge_dir / "responses"

    @property
    def heartbeat_file(self) -> Path:
        return self.config.bridge_dir / "heartbeat.txt"

    def is_alive(self) -> bool:
        """True if the Lua bridge has touched its heartbeat file recently."""
        try:
            mtime = self.heartbeat_file.stat().st_mtime
        except FileNotFoundError:
            return False
        return (time.time() - mtime) < HEARTBEAT_STALE_AFTER_SEC

    def probe(self) -> bool:
        return self.is_alive()

    def close(self) -> None:
        pass  # no persistent connection to release

    # -- request/response ------------------------------------------------------

    def call(self, op: str, args: dict | None = None, retries: int = 0, timeout: float | None = None) -> dict:
        last_exc: Exception | None = None
        for attempt in range(retries + 1):
            try:
                return self._call_once(op, args or {}, timeout)
            except BridgeError as exc:
                last_exc = exc
                if attempt < retries:
                    time.sleep(0.2)
                    continue
        assert last_exc is not None
        raise last_exc

    def _call_once(self, op: str, args: dict, timeout: float | None = None) -> dict:
        if not self.is_alive():
            raise BridgeNotConnected(
                f"REAPER bridge heartbeat not found or stale at {self.heartbeat_file}. "
                "Is REAPER running with reaper_bridge.lua loaded (Actions -> Show action "
                "list -> run reaper_bridge.lua)? Run the reaper_status tool for diagnostics."
            )

        self.requests_dir.mkdir(parents=True, exist_ok=True)
        req_id = next(self._id_counter)
        payload = json.dumps({"id": req_id, "op": op, "args": args})
        self._write_atomic(self.requests_dir / f"req_{req_id}.json", payload)

        response_path = self.responses_dir / f"resp_{req_id}.json"
        effective_timeout = timeout if timeout is not None else self.config.request_timeout
        deadline = time.monotonic() + effective_timeout
        while time.monotonic() < deadline:
            if response_path.exists():
                try:
                    content = response_path.read_text(encoding="utf-8")
                except OSError:
                    time.sleep(self.config.poll_interval)
                    continue
                response_path.unlink(missing_ok=True)
                try:
                    msg = json.loads(content)
                except json.JSONDecodeError as exc:
                    raise BridgeError(f"malformed response from bridge: {content!r}") from exc
                if not msg.get("ok", False):
                    raise BridgeError(f"REAPER bridge error for op '{op}': {msg.get('error')}")
                return msg.get("result", {})
            time.sleep(self.config.poll_interval)

        raise BridgeError(f"timed out waiting for bridge response to op '{op}' (id={req_id})")

    @staticmethod
    def _write_atomic(path: Path, content: str) -> None:
        fd, tmp_name = tempfile.mkstemp(dir=str(path.parent), prefix=path.name + ".", suffix=".tmp")
        try:
            with os.fdopen(fd, "w", encoding="utf-8") as f:
                f.write(content)
            os.replace(tmp_name, path)
        except BaseException:
            if os.path.exists(tmp_name):
                os.remove(tmp_name)
            raise


_default_client: BridgeClient | None = None


def get_default_client() -> BridgeClient:
    global _default_client
    if _default_client is None:
        _default_client = BridgeClient()
    return _default_client

File: crates/dawmcp-reaper/src/bridge_client.rs
//! File-based IPC client for `lua/reaper_bridge.lua`, ported from the
//! Python `bridge_client.py` it replaces. Protocol is unchanged: one JSON
//! request file per call, poll for the matching response file, clean both up.
//! See `lua/reaper_bridge.lua` for the Lua side and `docs/ARCHITECTURE.md`
//! for why this is file IPC and not a socket.

use std::path::{Path, PathBuf};
use std::time::{Duration, Instant, SystemTime};

use anyhow::{anyhow, bail, Context, Result};
use serde_json::Value;
use tokio::sync::Mutex;

const HEARTBEAT_STALE_AFTER: Duration = Duration::from_millis(2000);
const DEFAULT_REQUEST_TIMEOUT: Duration = Duration::from_secs(5);
const POLL_INTERVAL: Duration = Duration::from_millis(10);

pub struct BridgeClient {
    bridge_dir: PathBuf,
    client_id: String,
    counter: Mutex<u64>,
    request_timeout: Duration,
}

impl BridgeClient {
    pub fn new(bridge_dir: PathBuf) -> Self {
        Self {
            bridge_dir,
            client_id: uuid::Uuid::new_v4().simple().to_string()[..8].to_string(),
            counter: Mutex::new(0),
            request_timeout: DEFAULT_REQUEST_TIMEOUT,
        }
    }

    fn requests_dir(&self) -> PathBuf {
        self.bridge_dir.join("requests")
    }

    fn responses_dir(&self) -> PathBuf {
        self.bridge_dir.join("responses")
    }

    fn heartbeat_file(&self) -> PathBuf {
        self.bridge_dir.join("heartbeat.txt")
    }

    #[cfg(test)]
    pub(crate) fn client_id(&self) -> &str {
        &self.client_id
    }

    /// True if `reaper_bridge.lua` has touched its heartbeat file recently.
    pub fn is_alive(&self) -> bool {
        let Ok(meta) = std::fs::metadata(self.heartbeat_file()) else {
            return false;
        };
        let Ok(modified) = meta.modified() else {
            return false;
        };
        SystemTime::now()
            .duration_since(modified)
            .map(|age| age < HEARTBEAT_STALE_AFTER)
            .unwrap_or(false)
    }

    pub async fn call(&self, op: &str, args: Value) -> Result<Value> {
        self.call_with_timeout(op, args, self.request_timeout).await
    }

    pub async fn call_with_timeout(&self, op: &str, args: Value, timeout: Duration) -> Result<Value> {
        if !self.is_alive() {
            bail!(
                "REAPER bridge heartbeat not found or stale at {}. Is REAPER running with \
                 reaper_bridge.lua loaded (Actions -> Show action list -> run reaper_bridge.lua)?",
                self.heartbeat_file().display()
            );
        }

        std::fs::create_dir_all(self.requests_dir()).context("creating bridge requests dir")?;

        let req_id = {
            let mut counter = self.counter.lock().await;
            *counter += 1;
            format!("{}-{}", self.client_id, *counter)
        };

        let payload = serde_json::json!({ "id": req_id, "op": op, "args": args }).to_string();
        write_atomic(&self.requests_dir().join(format!("req_{req_id}.json")), &payload)
            .context("writing bridge request file")?;

        let response_path = self.responses_dir().join(format!("resp_{req_id}.json"));
        let deadline = Instant::now() + timeout;
        loop {
            if response_path.exists() {
                let content = match std::fs::read_to_string(&response_path) {
                    Ok(c) => c,
                    Err(_) => {
                        tokio::time::sleep(POLL_INTERVAL).await;
                        continue;
                    }
                };
                let _ = std::fs::remove_file(&response_path);
                let msg: Value = serde_json::from_str(&content)
                    .with_context(|| format!("malformed response from bridge: {content:?}"))?;
                let ok = msg.get("ok").and_then(Value::as_bool).unwrap_or(false);
                if !ok {
                    let err = msg.get("error").cloned().unwrap_or(Value::Null);
                    return Err(anyhow!("REAPER bridge error for op '{op}': {err}"));
                }
                return Ok(msg.get("result").cloned().unwrap_or(Value::Object(Default::default())));
            }
            if Instant::now() >= deadline {
                bail!("timed out waiting for bridge response to op '{op}' (id={req_id})");
            }
            tokio::time::sleep(POLL_INTERVAL).await;
        }
    }
}

fn write_atomic(path: &Path, content: &str) -> Result<()> {
    let tmp_path = path.with_extension("tmp");
    std::fs::write(&tmp_path, content)?;
    std::fs::rename(&tmp_path, path)?;
    Ok(())
}

/// Resolves the bridge IPC directory, matching `default_bridge_dir()` in the
/// Python `bridge_client.py`: `REAPER_MCP_BRIDGE_DIR` env override first,
/// else the first discovered REAPER install's `Scripts/reaper_mcp_bridge`,
/// else a last-resort fallback so callers still get a stable path to report
/// in errors.
pub fn default_bridge_dir() -> PathBuf {
    if let Ok(dir) = std::env::var("REAPER_MCP_BRIDGE_DIR") {
        return PathBuf::from(dir);
    }
    if let Some(install) = crate::discovery::find_reaper_installs().into_iter().next() {
        return PathBuf::from(install.scripts_dir).join("reaper_mcp_bridge");
    }
    dirs_fallback().join(".reaper_mcp_bridge")
}

fn dirs_fallback() -> PathBuf {
    std::env::var_os("USERPROFILE")
        .or_else(|| std::env::var_os("HOME"))
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("."))
}

#[cfg(test)]
mod tests {
    //! Ported from `tests/test_bridge_client.py`: a background task stands
    //! in for `reaper_bridge.lua`'s defer-loop pump, answering known ops and
    //! keeping the heartbeat fresh, so these exercise the real file-IPC
    //! protocol without a running REAPER.

    use super::*;
    use std::sync::atomic::{AtomicBool, Ordering};
    use std::sync::Arc;

    struct MockBridgeWorker {
        stop: Arc<AtomicBool>,
        handle: Option<tokio::task::JoinHandle<()>>,
    }

    impl MockBridgeWorker {
        fn start(bridge_dir: PathBuf) -> Self {
            let requests_dir = bridge_dir.join("requests");
            let responses_dir = bridge_dir.join("responses");
            let heartbeat_file = bridge_dir.join("heartbeat.txt");
            std::fs::create_dir_all(&requests_dir).unwrap();
            std::fs::create_dir_all(&responses_dir).unwrap();

            let stop = Arc::new(AtomicBool::new(false));
            let stop_clone = stop.clone();
            let handle = tokio::spawn(async move {
                while !stop_clone.load(Ordering::Relaxed) {
                    let _ = std::fs::write(&heartbeat_file, "tick");
                    if let Ok(entries) = std::fs::read_dir(&requests_dir) {
                        for entry in entries.flatten() {
                            let path = entry.path();
                            let Some(name) = path.file_name().and_then(|n| n.to_str()) else { continue };
                            if !name.starts_with("req_") {
                                continue;
                            }
                            let Ok(content) = std::fs::read_to_string(&path) else { continue };
                            let _ = std::fs::remove_file(&path);
                            let Ok(req): Result<Value, _> = serde_json::from_str(&content) else { continue };
                            let id = req.get("id").and_then(|v| v.as_str()).unwrap_or_default();
                            let op = req.get("op").and_then(|v| v.as_str()).unwrap_or_default();
                            let resp = if op == "ping" {
                                serde_json::json!({ "id": id, "ok": true, "result": { "pong": true } })
                            } else {
                                serde_json::json!({ "id": id, "ok": false, "error": format!("unknown op {op}") })
                            };
                            let resp_path = responses_dir.join(format!("resp_{id}.json"));
                            let _ = std::fs::write(&resp_path, resp.to_string());
                        }
                    }
                    tokio::time::sleep(Duration::from_millis(5)).await;
                }
            });
            Self { stop, handle: Some(handle) }
        }
    }

    impl Drop for MockBridgeWorker {
        fn drop(&mut self) {
            self.stop.store(true, Ordering::Relaxed);
            if let Some(handle) = self.handle.take() {
                handle.abort();
            }
        }
    }

    #[tokio::test]
    async fn call_success() {
        let dir = tempfile::tempdir().unwrap();
        let _worker = MockBridgeWorker::start(dir.path().to_path_buf());
        tokio::time::sleep(Duration::from_millis(20)).await;

        let client = BridgeClient::new(dir.path().to_path_buf());
        let result = client.call("ping", serde_json::json!({})).await.unwrap();
        assert_eq!(result, serde_json::json!({ "pong": true }));
    }

    #[tokio::test]
    async fn call_unknown_op_raises() {
        let dir = tempfile::tempdir().unwrap();
        let _worker = MockBridgeWorker::start(dir.path().to_path_buf());
        tokio::time::sleep(Duration::from_millis(20)).await;

        let client = BridgeClient::new(dir.path().to_path_buf());
        let err = client.call("nonexistent_op", serde_json::json!({})).await.unwrap_err();
        assert!(err.to_string().contains("unknown op"));
    }

    #[tokio::test]
    async fn no_heartbeat_raises() {
        let dir = tempfile::tempdir().unwrap();
        let client = BridgeClient::new(dir.path().join("no_bridge_here"));
        let err = client.call("ping", serde_json::json!({})).await.unwrap_err();
        assert!(err.to_string().contains("heartbeat"));
    }

    #[tokio::test]
    async fn is_alive_reflects_heartbeat_freshness() {
        let dir = tempfile::tempdir().unwrap();
        let worker = MockBridgeWorker::start(dir.path().to_path_buf());
        tokio::time::sleep(Duration::from_millis(20)).await;

        let client = BridgeClient::new(dir.path().to_path_buf());
        assert!(client.is_alive());

        drop(worker);
        let dead_client = BridgeClient::new(dir.path().join("dead"));
        assert!(!dead_client.is_alive());
    }

    #[tokio::test]
    async fn concurrent_clients_do_not_collide_on_request_ids() {
        let dir = tempfile::tempdir().unwrap();
        let _worker = MockBridgeWorker::start(dir.path().to_path_buf());
        tokio::time::sleep(Duration::from_millis(20)).await;

        let client_a = BridgeClient::new(dir.path().to_path_buf());
        let client_b = BridgeClient::new(dir.path().to_path_buf());
        assert_ne!(client_a.client_id(), client_b.client_id());

        let result_a = client_a.call("ping", serde_json::json!({})).await.unwrap();
        let result_b = client_b.call("ping", serde_json::json!({})).await.unwrap();
        assert_eq!(result_a, serde_json::json!({ "pong": true }));
        assert_eq!(result_b, serde_json::json!({ "pong": true }));
    }
}
